File: jpress-engine/src/main/java/by/ibragimov/jpress/utils/FileService.java
package by.ibragimov.jpress.utils;

import org.apache.commons.io.FileUtils;
import org.apache.commons.lang3.tuple.Pair;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.io.File;
import java.io.IOException;
import java.nio.file.*;
import java.nio.file.attribute.BasicFileAttributes;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * Contains file processing helpers.
 *
 * I decide to make all utils plain spring bean to simplify testing and overall consistency.
 *
 * @author Ibragimov Ruslan
 */
@Service
public class FileService {

    private static final Logger LOGGER = LoggerFactory.getLogger(FileService.class);

    /**
     * Copy all files from path
     *
     * @param path
     * @return
     */
    public boolean copyFolder(final String path) {
        return false;
    }

    /**
     * Return list of {@link Path}s that contains all files in provided directory and subdirectories.
     *
     * @param folder directory from which we starting scan.
     * @return list of {@link Pair}s: left in pair - folder, right - file.
     */
    public List<Pair<Path, Path>> scanFolder(Path folder) {
        List<Pair<Path, Path>> paths = new ArrayList<>();
        try {
            Files.list(folder).forEach(path -> {
                if (Files.isDirectory(path)) {
                    paths.addAll(scanFolder(path));
                } else if (Files.isRegularFile(path)) {
                    paths.add(Pair.of(folder, path));
                } else {
                    LOGGER.error("Hmmmm, '{}'", path.toAbsolutePath());
                }
            });
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        return paths;
    }
}

File: kpress-engine/src/main/kotlin/io/heapy/kpress/fs/FileService.kt
package io.heapy.kpress.fs

import io.heapy.kpress.extensions.logger
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.util.ArrayList

/**
 * Contains file processing helpers.
 *
 * I decide to make all utils plain spring bean to simplify testing and overall consistency.
 *
 * @author Ruslan Ibragimov
 */
class FileService {

    /**
     * Copy all files from path
     *
     * @param path
     * @return
     */
    fun copyFolder(path: String): Boolean {
        return false
    }

    /**
     * Return list of [Path]s that contains all files in provided directory and subdirectories.
     *
     * @param folder directory from which we starting scan.
     * @return list of [Pair]s: left in pair - folder, right - file.
     */
    fun scanFolder(folder: Path): List<Pair<Path, Path>> {
        val paths = ArrayList<Pair<Path, Path>>()
        try {
            Files.list(folder).forEach { path ->
                when {
                    Files.isDirectory(path) -> paths.addAll(scanFolder(path))
                    Files.isRegularFile(path) -> paths.add(Pair(folder, path))
                    else -> LOGGER.error("Hmmmm, '{}'", path.toAbsolutePath())
                }
            }
        } catch (e: IOException) {
            throw RuntimeException(e)
        }

        return paths
    }

    companion object {
        private val LOGGER = logger<FileService>()
    }
}
